File: radar_lassez/nodes/deduplicator.ts
import stringSimilarity from 'string-similarity';
import { prisma } from '../lib/prisma';
import { IngestedArticle } from './ingestion';
import { getEffectiveParam } from '../lib/config-resolver';

export interface MergedTopic {
    clusterTitle: string;
    articles: IngestedArticle[];
    aggregatedBias: string[];
    date: Date;
}

/**
 * Nœud 2 : Deduplicator (Tamisage & Dédoublonnage Algorithmique)
 * 
 * Regroupe les articles par similarité sémantique de titre (Dice's Coefficient).
 * Compare les clusters formés avec l'historique de la base de données (48h par défaut)
 * pour ne laisser passer que les nouveaux sujets d'investigation originaux.
 */
export async function runDeduplicatorNode(articles: IngestedArticle[]) {
    console.log(`[Node 2: Deduplicator] 🧩 Démarrage du tamisage sur ${articles.length} articles bruts.`);
    
    if (!articles || articles.length === 0) return;

    // Récupération dynamique du seuil autorisé (Cascade)
    const threshold = await getEffectiveParam('dedup', 'similarityThreshold', 0.45);

    const clusters: MergedTopic[] = [];

    // 1. Groupement algorithmique (Clustering/Deduplication)
    for (const article of articles) {
        let foundCluster = false;

        for (const cluster of clusters) {
            const similarity = stringSimilarity.compareTwoStrings(
                (article.title || '').toLowerCase(),
                (cluster.clusterTitle || '').toLowerCase()
            );

            if (similarity >= threshold) {
                cluster.articles.push(article);
                
                if (article.source_bias && !cluster.aggregatedBias.includes(article.source_bias)) {
                    cluster.aggregatedBias.push(article.source_bias);
                }
                
                foundCluster = true;
                break;
            }
        }

        if (!foundCluster) {
            clusters.push({
                clusterTitle: article.title,
                articles: [article],
                aggregatedBias: article.source_bias ? [article.source_bias] : ['Indépendant'],
                date: article.pubDate || new Date()
            });
        }
    }

    console.log(`[Node 2: Deduplicator] 📉 Réduction intrajournalière : ${articles.length} articles compilés en ${clusters.length} sujets synthétiques.`);
    
    // 2. Dé-duplication HISTORIQUE absolue (Vérification en DB)
    console.log(`[Node 2: Deduplicator] 🕵️‍♂️ Vérification de l'historique en base de données...`);
    
    const lookbackHours = await getEffectiveParam('dedup', 'dedupLookbackHours', 48);
    const historyCutoffDate = new Date(Date.now() - (lookbackHours * 60 * 60 * 1000));

    const historicalTopics = await prisma.newsTopic.findMany({
        where: {
            createdAt: { gte: historyCutoffDate }
        },
        select: { raw_data: true }
    });

    const historicalTitles: string[] = [];
    for (const h of historicalTopics) {
        try {
            const parsed = JSON.parse(h.raw_data);
            if (parsed && parsed.clusterTitle) {
                historicalTitles.push(parsed.clusterTitle.toLowerCase());
            }
        } catch (e) { }
    }

    let savedCount = 0;
    let ignoredCount = 0;

    // 3. Persistance dans la base de données (NewsTopic)
    const topicsToCreate = [];

    for (const cluster of clusters) {
        try {
            const currentTitleLower = cluster.clusterTitle.toLowerCase();
            let isDuplicateHistory = false;

            for (const hTitle of historicalTitles) {
                const histSim = stringSimilarity.compareTwoStrings(currentTitleLower, hTitle);
                if (histSim >= (threshold * 0.8)) {
                    isDuplicateHistory = true;
                    break;
                }
            }

            if (isDuplicateHistory) {
                ignoredCount++;
                continue;
            }

            topicsToCreate.push({
                raw_data: JSON.stringify(cluster),
                status: 'INGESTED',
                tags: "[]"
            });
            savedCount++;
        } catch (error) {
            console.error(`[Node 2] ❌ Erreur lors de la préparation pour le topic "${cluster.clusterTitle}":`, error);
        }
    }

    if (topicsToCreate.length > 0) {
        try {
            await prisma.newsTopic.createMany({
                data: topicsToCreate
            });
        } catch (error) {
            console.error(`[Node 2] ❌ Erreur lors de l'insertion DB en masse :`, error);
        }
    }

    console.log(`[Node 2: Deduplicator] ✅ Fin du tamis. ${savedCount} nouveaux NewsTopics injectés. (${ignoredCount} doublons historiques rejetés).`);
}

File: daemon/internal/nodes/deduplicator.go
package nodes

import (
	"encoding/json"
	"log"
	"strings"
	"time"

	"github.com/Flayrox/LASSEZ/daemon/internal/config"
	"github.com/Flayrox/LASSEZ/daemon/internal/payload"
)

// mergedTopic mirrors the TS MergedTopic persisted in raw_data.
type mergedTopic struct {
	ClusterTitle   string            `json:"clusterTitle"`
	Articles       []IngestedArticle `json:"articles"`
	AggregatedBias []string          `json:"aggregatedBias"`
	Date           time.Time         `json:"date"`
}

// RunDeduplicator is Node 2 of the pipeline: clusters the ingested articles
// by title similarity, drops historical duplicates and persists the new
// topics as INGESTED signals.
func RunDeduplicator(client *payload.Client, resolver *config.Resolver, articles []IngestedArticle) error {
	log.Printf("[Node 2: Deduplicator] 🧩 Tamisage sur %d articles bruts.", len(articles))
	if len(articles) == 0 {
		return nil
	}

	threshold := 0.45
	if v := resolver.GetEffectiveParam("dedup", "similarityThreshold", 0.45); v != nil {
		threshold = toFloat64(v, 0.45)
	}

	// 1. Clustering by title similarity.
	var clusters []mergedTopic
	for _, article := range articles {
		found := false
		for i := range clusters {
			if compareTwoStrings(strings.ToLower(article.Title), strings.ToLower(clusters[i].ClusterTitle)) >= threshold {
				clusters[i].Articles = append(clusters[i].Articles, article)
				if article.SourceBias != "" && !containsString(clusters[i].AggregatedBias, article.SourceBias) {
					clusters[i].AggregatedBias = append(clusters[i].AggregatedBias, article.SourceBias)
				}
				found = true
				break
			}
		}
		if !found {
			bias := []string{"Indépendant"}
			if article.SourceBias != "" {
				bias = []string{article.SourceBias}
			}
			date := article.PubDate
			if date.IsZero() {
				date = time.Now()
			}
			clusters = append(clusters, mergedTopic{
				ClusterTitle:   article.Title,
				Articles:       []IngestedArticle{article},
				AggregatedBias: bias,
				Date:           date,
			})
		}
	}

	log.Printf("[Node 2: Deduplicator] 📉 %d articles compilés en %d sujets.", len(articles), len(clusters))

	// 2. Historical deduplication against existing signals.
	lookbackHours := 48.0
	if v := resolver.GetEffectiveParam("dedup", "dedupLookbackHours", float64(48)); v != nil {
		lookbackHours = toFloat64(v, 48)
	}
	cutoff := time.Now().Add(-time.Duration(lookbackHours) * time.Hour)

	historicalTitles := []string{}
	if historical, err := client.GetSignalsSince(cutoff); err != nil {
		log.Printf("[Node 2] ❌ Lecture historique: %v", err)
	} else {
		for _, h := range historical {
			var raw struct {
				ClusterTitle string `json:"clusterTitle"`
			}
			if err := json.Unmarshal(h.RawData, &raw); err == nil && raw.ClusterTitle != "" {
				historicalTitles = append(historicalTitles, strings.ToLower(raw.ClusterTitle))
			}
		}
	}

	// 3. Persist new topics.
	var toCreate []map[string]any
	saved, ignored := 0, 0
	for _, cluster := range clusters {
		current := strings.ToLower(cluster.ClusterTitle)
		duplicate := false
		for _, h := range historicalTitles {
			if compareTwoStrings(current, h) >= threshold*0.8 {
				duplicate = true
				break
			}
		}
		if duplicate {
			ignored++
			continue
		}

		rawJSON, err := json.Marshal(cluster)
		if err != nil {
			continue
		}
		toCreate = append(toCreate, map[string]any{
			"raw_data": string(rawJSON),
			"status":   "INGESTED",
			"tags":     "[]",
		})
		saved++
	}

	if len(toCreate) > 0 {
		if err := client.CreateSignals(toCreate); err != nil {
			log.Printf("[Node 2] ❌ Insertion Payload: %v", err)
		}
	}

	log.Printf("[Node 2: Deduplicator] ✅ %d Signals injectés (%d doublons historiques rejetés).", saved, ignored)
	return nil
}
